src/mtl.rs

```diff
@@ -79,50 +111,52 @@ impl<'a> From<Material> for Cow<'a, Material> {
 struct Parser<I>(I);
 
 impl<'a, I: Iterator<Item = &'a str>> Parser<I> {
-    fn get_vec(&mut self) -> Option<[f32; 3]> {
+    fn get_vec(&mut self) -> Result<[f32; 3], MtlError> {
         let (x, y, z) = match (self.0.next(), self.0.next(), self.0.next()) {
             (Some(x), Some(y), Some(z)) => (x, y, z),
             other => {
-                println!("invalid {:?}", other);
-                return None;
+                return Err(MtlError::InvalidValue(format!("{:?}", other)));
             }
         };
 
         match (x.parse::<f32>(), y.parse::<f32>(), z.parse::<f32>()) {
-            (Ok(x), Ok(y), Ok(z)) => Some([x, y, z]),
+            (Ok(x), Ok(y), Ok(z)) => Ok([x, y, z]),
             other => {
-                println!("invalid {:?}", other);
-                None
+                Err(MtlError::InvalidValue(format!("{:?}", other)))
             }
         }
     }
 
-    fn get_i32(&mut self) -> Option<i32> {
+    fn get_i32(&mut self) -> Result<i32, MtlError> {
         match self.0.next() {
-            Some(v) => FromStr::from_str(v).ok(),
+            Some(v) => FromStr::from_str(v).map_err(|_| MtlError::InvalidValue(v.to_string())),
             None => {
-                println!("missing i32");
-                None
+                Err(MtlError::MissingValue(MtlMissingType::I32))
             }
         }
     }
 
-    fn get_f32(&mut self) -> Option<f32> {
+    fn get_f32(&mut self) -> Result<f32, MtlError> {
         match self.0.next() {
-            Some(v) => FromStr::from_str(v).ok(),
+            Some(v) => FromStr::from_str(v).map_err(|_| MtlError::InvalidValue(v.to_string())),
             None => {
-                println!("missing f32");
-                None
+                Err(MtlError::MissingValue(MtlMissingType::F32))
             }
         }
     }
 
-    fn get_string(&mut self) -> Option<String> {
+    fn get_string(mut self) -> Result<String, MtlError> {
```

Comment:
hmm, this is still inconsistent with the other methods.
If you need it to have `self` by value, need to rename it to `into_string`

Reply:
Fixed.